File: C++/22862.cpp
#include <iostream>
#include <string>
#include <algorithm>

using namespace std;

int main(){
    int n = 0;
    int k;
    int l[1000000];
    int odd = 0; 
    int tot = 0; 
    int ans = 0; 
    int right = 0;

    cin >> n >> k;
    for(int i = 0; i < n; i++){
        cin >> l[i];
    }


    for(int i = 0; i < n; i++){

        if(l[i]%2 == 0){
            tot++;
        }else{
            if( odd+1 > k){
                ans = max(ans, tot);
                right = i;
                break;
            }else{
                odd++;
            }
        }

        if(odd < k){
            ans = max(ans, i-odd+1);
        }
    }

    for(int i = 0; i < n-right; i++){
        if (l[i] % 2 == 0) {
            tot--;
        }
        else{
            odd--;
        }
        while(right < n){
            if(l[right]%2 == 0){
                tot++;
            }else{
                if(odd+1 > k){
                    ans = max(ans, tot);
                    break;
                }else{
                    odd++;
                }
            }
            right ++;
        }
        if (odd < k){
            ans = max(ans, (right-i-odd));
        }
    }

    cout << ans;

    return 0;
}

File: C++/22862.py
# 10:55 ~ 
#
# N = 100만 -> NlogN / N 까지만 가능하다
#
# 투포인터를 정하고, 짝수, 홀수를 따로 관리한다.
# 즉, 홀수가 coun가 k 를 넘는자면 멈춘다. 
#
# 양끝 홀수가 반례 같지만.. 어차피 sliding window 로 탐색한다면 다른 케이스까지 고려된다
#
# 반례) 모든게 다 이미 짝수라면


n,k = map(int, input().split())

l = list(map(int, input().split()))

odd = 0
total = 0

answer = 0
right = 0

for a in range(len(l)):
    if l[a] % 2 == 0:
        total += 1            
    else:
        if odd+1 > k:
            answer = max(answer, total)
            right = a
            break
        else:
            odd += 1
            
    if odd < k:
        answer = max(answer, (a-odd+1))
            
for left in range(len(l)-right):

    if l[left] % 2 ==0:
        total -= 1
    else:
        odd -= 1
    
    while right < len(l):
        if l[right]%2 == 0:
            total += 1
        else:
            if odd+1 > k:
                answer = max(answer, total)
                break
            else:
                odd += 1
        right += 1
    
    if odd < k:
        answer = max(answer, (right-left-odd))
    
print(answer)
        